sdk/core/Azure.Core/src/Pipeline/Internal/ClientRequestIdPolicy.cs

```diff
@@ -3,6 +3,9 @@
 
 namespace Azure.Core.Pipeline
 {
+#if NET6_0_OR_GREATER
+    [System.Diagnostics.CodeAnalysis.DynamicallyAccessedMembers(System.Diagnostics.CodeAnalysis.DynamicallyAccessedMemberTypes.PublicMethods)]
```

Comment:
I'm not sure this is doing what you intend. What warning you are trying to address with applying `DynamicallyAccessedMembers` on this `class`?

See https://github.com/dotnet/runtime/issues/49465 for more information on which scenarios this supports. @vitek-karas - do we have official docs on this?

Reply:
Not really - we need to add them. This will only work if there's a `instance.GetType()` call where `instance` is statically typed as `ClientRequestIdPolicy` - without the `GetType` call the attribute does nothing. And it's certainly not a hint to the trimmer that it should always preserved the members defined in the attribute.

Reply:
This resolves a warning from this line - https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/Pipeline/HttpPipelineSynchronousPolicy.cs#L28

Reply:
If I remember correctly, it was an IL2026 warning. The warning was coming from the reflection code linked above, where the abstract class constructor is checking if the inherited class has overridden a virtual method. So my intention was to preserve the public methods on the inherited types. I wasn't 100% sure if this was the way to do it, but after adding the attributes the warning went away.

Reply:
> This resolves a warning from this line - https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/Pipeline/HttpPipelineSynchronousPolicy.cs#L28

I would not expect that to resolve this warning - I think it's the attribute on `HttpPipelineSynchronousPolicy` that is fixing the warning. It shouldn't be necessary on this class.

Note that the reflection pattern there (trying to figure out if a method was overriden) will not reliably work in general. It is not possible to answer this question in C#/reflection APIs reliably. Not because of trimming or AOT, it's just impossible in general. I know this because we have a similar need in the base class libraries and it is fixed there by special casing within the runtime: https://github.com/dotnet/runtime/blob/e2b5936b758c10fe4f4f287393cb14a288febbd7/src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs#L452-L458. There's no guarantee an override will have the same name as the base method. One cannot look for it like this.

This could also be fixed by rolling back the PR that introduced it - per https://github.com/Azure/azure-sdk-for-net/pull/18197#issuecomment-768417154 this was a perf improvement that lead to a "Close to noise improvement in the perf benchmark we have.", but introduced a correctness issue. I would suggest rolling that back in a separate PR instead of annotating this.

You'd get much better trimming without having to attribute this class.

Reply:
Just to give an example of why it's wrong to help getting the revert in:

```csharp
using System;
using System.Reflection;

// Prints True
new Derived1();

// Prints False
new Derived2();

class Base
{
    public Base()
    {
        var onReceivedResponseMethod = GetType().GetMethod(nameof(DoSomething), BindingFlags.Instance | BindingFlags.Public, null, Type.EmptyTypes, null);
        if (onReceivedResponseMethod != null)
        {
            bool overriden = onReceivedResponseMethod.GetBaseDefinition().DeclaringType != onReceivedResponseMethod.DeclaringType;
            Console.WriteLine($"{GetType()}: {overriden}");
        }
    }

    public virtual object DoSomething() => null;
}

class Derived1 : Base
{
    public override object DoSomething() => null;
}

class Derived2 : Base
{
    public override string DoSomething() => null;
}
```

Reply:
I removed these extra attributes, and the warning is still one. At some point during my testing, I had thought they were necessary in all places.
In terms of reverting this change, I understand what you're saying about this being incorrect. I'll remove the attribute and work on some perf testing and getting this reverted in a separate PR.

Reply:
@MichalStrehovsky after discussing this a bit internally, the consensus was that we'd like to keep the perf improvement+attribute. The `OnReceivedResponse` method has the following signature:
```
public virtual void OnReceivedResponse(HttpMessage message) { }
```
Since it returns void, and HttpMessage is a sealed class, this specific case is compatible with using reflection in this way to see if it's been overridden.

Reply:
> Since it returns void, and HttpMessage is a sealed class, this specific case is compatible with using reflection in this way to see if it's been overridden.

The example with covariant returns in C# was just an easy-to-give example.

This form of override detection would already be broken 20 years ago (before covariant returns) because the general mechanism to override method with another method that doesn't have a matching name has existed for 20+ years in the .NET runtime and file format. C# just hasn't been using it for non-interface virtual methods until covariant returns were introduced so that's where I can show a C# example easily.

The pattern is probably fine/safe for C# today. I don't know if it's fine/safe for non-C# languages like VB, F#, or C++/CLI. I don't know it it will be safe for C# in the future. As a runtime dev I could give you an example in IL language where it would be broken today. We have a capability in the runtime for methods with different names to override other methods and we do test it. It is up to language developers to use it as they need.

I would recommend you to reconsider, especially if the perf improvement was miniscule. It is a correctness issue.

There's no reflection API to read this information (it is tracked in https://github.com/dotnet/runtime/issues/81299). It's impossible to write this with reflection correctly today.

Reply:
The performance analysis is a bit too expensive to do right now, since there cannot be any noticeable performance regression here. I opened an issue https://github.com/Azure/azure-sdk-for-net/issues/38225 to track removing this attribute when we have resources available to thoroughly test the impact this change would have on performance.